src/libraries/System.Diagnostics.Process/tests/ProcessTests.cs

```diff
@@ -1556,75 +1556,75 @@ public void MainWindowHandle_GetNotStarted_ThrowsInvalidOperationException()
-            {
-                try
-                {
-                    for (int attempt = 0; attempt < 50; ++attempt)
-                    {
-                        process.Refresh();
-                        if (process.MainWindowHandle != IntPtr.Zero)
-                        {
-                            break;
-                        }
-
-                        Thread.Sleep(100);
-                    }
+            VerifyProcessPublicPropertyAndPrivateFieldAfterRefresh(process => process.MainWindowHandle, IntPtr.Zero, "_haveMainWindow", false);
+        }
 
-                    Assert.NotEqual(IntPtr.Zero, process.MainWindowHandle);
-                }
-                finally
-                {
-                    process.Kill();
-                    Assert.True(process.WaitForExit(WaitInMS));
-                }
-            }
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNotWindowsNanoServer))]
+        [OuterLoop("Launches notepad")]
+        [PlatformSpecific(TestPlatforms.Windows)]
+        public void MainWindowTitle_GetsRefreshedAfterEveryRefresh()
+        {
+            VerifyProcessPublicPropertyAndPrivateFieldAfterRefresh<string, string>(process => process.MainWindowTitle, string.Empty, "_mainWindowTitle", null);
         }
 
-        [Fact]
-        [OuterLoop]
-        [Trait(XunitConstants.Category, XunitConstants.IgnoreForCI)] // Pops UI
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNotWindowsNanoServer))]
+        [OuterLoop("Launches notepad")]
         [PlatformSpecific(TestPlatforms.Windows)]
-        public void MainWindowTitle_GetWithGui_ShouldRefresh_Windows()
+        public void IsResponding_GetsRefreshedAfterEveryRefresh()
+        {
+            VerifyProcessPublicPropertyAndPrivateFieldAfterRefresh(process => process.Responding, false, "_haveResponding", false);
+        }
+
+        private void VerifyProcessPublicPropertyAndPrivateFieldAfterRefresh<TPublic, TPrivate>(
+            Func<Process, TPublic> publicPropertyGetter, TPublic defaultPublicPropertyValue,
+            string privateFieldName, TPrivate defaultPrivateFieldValue)
         {
             const string ExePath = "notepad.exe";
-            Assert.True(IsProgramInstalled(ExePath));
+            Assert.True(IsProgramInstalled(ExePath), "Notepad is not installed");
 
-            using (Process process = Process.Start(ExePath))
+            using (Process process = Process.Start(new ProcessStartInfo(ExePath)))
             {
-                try
+                int triesRemaining = 50;
+                while (process.MainWindowHandle == IntPtr.Zero) // the process was started, but main Window was not loaded yet
                 {
-                    Assert.Equal(string.Empty, process.MainWindowTitle);
+                    Thread.Sleep(100);
+
+                    Assert.True(--triesRemaining > 0, "Notepad has failed to load for 5s, something is wrong");
```

Comment:
Fine if it works for you, but normally I'm super pessimistic about timeouts as I've seen stuff be incredibly slow occasionally. Eg., 30 secs

Reply:
I don't expect it to take more than 200ms, but to be 100% sure that we don't give up too early I've set it to 5s. Perhaps it's too long and 1s would be enough?